{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kzxBre5T73cg"
      },
      "source": [
        "# ML4NGP training school 2026\n",
        "\n",
        "# Deep generative models for IDP structural ensembles üß¨\n",
        "\n",
        "# Hands-on session\n",
        "\n",
        "---\n",
        "\n",
        "## From sequence to coarse-grained structural ensembles with idpGAN\n",
        "\n",
        "Welcome!\n",
        "\n",
        "In this session, we will explore **deep generative models** that we can use to obtain structural ensembles of intrinsically disordered proteins (IDPs).\n",
        "\n",
        "## Generative modeling\n",
        "\n",
        "We will first use **[idpGAN](https://www.nature.com/articles/s41467-023-36443-x)**, a first-generation generative model trained to learn the distribution of conformations of coarse-grained (CG) IDPs.\n",
        "\n",
        "At the end of the notebook, we will analyze some ensembles generated with more recent ML ensemble generators.\n",
        "\n",
        "## Ensemble analysis\n",
        "\n",
        "Generating an ensemble is only the first step. Another key part is their **analysis**.\n",
        "\n",
        "We will mainly use **[IDPEnsembleTools (IDPET)](https://onlinelibrary.wiley.com/doi/10.1002/pro.70427)**, a Python library designed specifically for analyzing IDP/IDR ensembles.\n",
        "\n",
        "We will also perform some analyses manually using **[NumPy](https://numpy.org/)**.\n",
        "\n",
        "## Outcomes\n",
        "\n",
        "By the end of this notebook, you will be able to:\n",
        "\n",
        "- ‚úÖ Generate a structural ensemble from a protein sequence with idpGAN\n",
        "- ‚úÖ Understand how ensemble data is represented in data analysis pipeline (numpy arrays or PyTorch tensors)\n",
        "- ‚úÖ Visualize some ensemble properties with IDPET (e.g., radius of gyration distributions)\n",
        "- ‚úÖ Compare structural ensembles (e.g.: to simulation-based references)\n",
        "- ‚úÖ Critically evaluate the strengths and limitations of ensemble generators"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MHNJVuS2Ho7H"
      },
      "source": [
        "# Part 1/6 - Install and setup the model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "F4eFKCQyHn98",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Install dependencies and define useful functions\n",
        "\n",
        "#@markdown Run this cell to install all the required dependencies for idpGAN. The installation will take place in the Colab notebook in the cloud. The whole process will take some minutes.\n",
        "\n",
        "#\n",
        "# Installation part.\n",
        "#\n",
        "\n",
        "# Download from GitHub the idpgan library and data.\n",
        "!if [ ! -f \"main.zip\" ]; then wget https://github.com/feiglab/idpgan/archive/refs/heads/main.zip; fi\n",
        "!if [ ! -d \"idpgan-main\" ]; then unzip -u main.zip idpgan-main/data/* idpgan-main/idpgan/*; fi\n",
        "!if [ ! -d \"data\" ]; then mv idpgan-main/data .; fi\n",
        "!if [ ! -d \"idpgan\" ]; then mv idpgan-main/idpgan .; fi\n",
        "# Install the python dependencies for analyzing and visualizing 3D conformations.\n",
        "!pip install mdtraj\n",
        "!pip install idpet\n",
        "!pip install py3Dmol\n",
        "\n",
        "\n",
        "#\n",
        "# Import the libraries we need.\n",
        "#\n",
        "\n",
        "import os\n",
        "import time\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import mdtraj\n",
        "import py3Dmol\n",
        "import torch\n",
        "\n",
        "from idpgan.data import (\n",
        "    parse_fasta_seq,\n",
        "    one_to_three,\n",
        ")\n",
        "from idpgan.nn_models import load_netg_article\n",
        "\n",
        "\n",
        "#\n",
        "# Useful functions for later.\n",
        "#\n",
        "\n",
        "def to_trajectory_and_top(xyz, seq, name, data_path):\n",
        "    top_path = os.path.join(data_path, f\"{name}.pdb\")\n",
        "    traj_path = os.path.join(data_path, f\"{name}.xtc\")\n",
        "    seq_to_cg_pdb(seq, out_fp=top_path)\n",
        "    pdb = mdtraj.load(top_path)\n",
        "    traj = mdtraj.Trajectory(xyz=xyz, topology=pdb.topology)\n",
        "    traj.save(traj_path)\n",
        "    return traj\n",
        "\n",
        "def seq_to_cg_pdb(seq, out_fp=None):\n",
        "    \"\"\"\n",
        "    Gets an amino acid sequence and returns a template\n",
        "    CG PDB file.\n",
        "    \"\"\"\n",
        "    pdb_lines = []\n",
        "    for i, aa_i in enumerate(seq):\n",
        "        res_idx = i + 1\n",
        "        aa_i = one_to_three[aa_i]\n",
        "        line_i = \"ATOM{:>7} CA   {} A{:>4}       0.000   0.000   0.000  1.00  0.00\\n\".format(\n",
        "                   str(res_idx), aa_i, str(res_idx))\n",
        "        pdb_lines.append(line_i)\n",
        "    pdb_content = \"\".join(pdb_lines)\n",
        "    if out_fp is not None:\n",
        "        with open(out_fp, \"w\") as o_fh:\n",
        "            o_fh.write(pdb_content)\n",
        "    return pdb_content\n",
        "\n",
        "def get_ca_topology(sequence: str):\n",
        "    topology = mdtraj.Topology()\n",
        "    chain = topology.add_chain()\n",
        "    for res in sequence:\n",
        "        res_obj = topology.add_residue(one_to_three[res], chain)\n",
        "        topology.add_atom(\"CA\", mdtraj.core.topology.elem.carbon, res_obj)\n",
        "    return topology\n",
        "\n",
        "def xyz_to_traj(xyz, seq):\n",
        "    top = get_ca_topology(seq)\n",
        "    traj = mdtraj.Trajectory(xyz=xyz, topology=top)\n",
        "    return traj\n",
        "\n",
        "\n",
        "def extract_rgb_from_colormap(colormap_name, N):\n",
        "    # Create a colormap object.\n",
        "    colormap = plt.get_cmap(colormap_name)\n",
        "    # Generate N equally spaced values between 0 and 1.\n",
        "    points = np.linspace(0, 1, N)\n",
        "    # Use the colormap to get colors for these points, converting to HTML format.\n",
        "    colors = [colormap(point)[:3] for point in points] # Extract RGB, ignore A.\n",
        "    html_colors = ['#' + ''.join([f'{int(val*255):02x}' for val in color]) for color in colors]\n",
        "    return html_colors\n",
        "\n",
        "def show_structure(\n",
        "        snapshot, representation='aa', cmap_name='viridis', max_structures=100\n",
        "    ):\n",
        "    pdb_fn = f'.temp.{time.time()}.pdb'\n",
        "    vtraj = mdtraj.Trajectory(\n",
        "        # max_structures only takes effect if we have more than 100 conformations.\n",
        "        xyz=snapshot.xyz[:max_structures], topology=snapshot.topology\n",
        "    )\n",
        "    vtraj.superpose(vtraj, 0)\n",
        "    vtraj.save(pdb_fn)\n",
        "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "    if len(vtraj) == 1:\n",
        "        with open(pdb_fn, 'r') as i_fh:\n",
        "            view.addModel(i_fh.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_fn, 'r') as i_fh:\n",
        "            view.addModelsAsFrames(\"\".join([x for x in i_fh]))\n",
        "    os.remove(pdb_fn)\n",
        "\n",
        "    if representation == 'aa':\n",
        "        view.setStyle({'cartoon': {'color': 'spectrum'},\n",
        "                       'stick': {'color': 'spectrum'}})\n",
        "    elif representation == 'cg':\n",
        "        n_residues = snapshot.xyz.shape[1]\n",
        "        colors = extract_rgb_from_colormap(cmap_name, n_residues)\n",
        "        for i in range(n_residues):\n",
        "            view.setStyle({'serial': str(i+1)},\n",
        "                          {'sphere': {'color': colors[i],\n",
        "                                      'radius': 1.9}})\n",
        "    else:\n",
        "        raise NotImplementedError(representation)\n",
        "\n",
        "    view.zoomTo()\n",
        "    if len(vtraj) > 1:\n",
        "        view.animate({'loop': \"forward\", 'reps': 1, 'interval':1000})\n",
        "    return view\n",
        "\n",
        "\n",
        "#\n",
        "# Setup some files.\n",
        "#\n",
        "\n",
        "temp_data_path = \"data\"\n",
        "\n",
        "for seq_name in (\"polyala\", \"protac\"):\n",
        "    _seq = parse_fasta_seq(os.path.join(temp_data_path, f\"{seq_name}.fasta\"))\n",
        "    seq_to_cg_pdb(\n",
        "        _seq, os.path.join(temp_data_path, f\"{seq_name}.pdb\")\n",
        "    )\n",
        "    _traj = mdtraj.Trajectory(\n",
        "        xyz=np.load(\n",
        "            os.path.join(temp_data_path, f\"{seq_name}.npy\")\n",
        "        ),\n",
        "        topology=mdtraj.load(\n",
        "            os.path.join(temp_data_path, f\"{seq_name}.pdb\")\n",
        "        ).topology\n",
        "    )\n",
        "    _traj[0:1000].save(os.path.join(temp_data_path, f\"{seq_name}.xtc\"))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0ZVy8w-kL_Ub"
      },
      "source": [
        "## IdpGAN discussion\n",
        "\n",
        "To understand the strengths and limitations of any AI model, you must know how it was trained. Before using a method, reading the original paper is always a good idea.\n",
        "\n",
        "Let‚Äôs briefly recap the key facts about **idpGAN**.\n",
        "\n",
        "### Model input/output\n",
        "**Input:** a protein sequence  \n",
        "**Output:** a 3D structural ensemble of a single IDP chain.\n",
        "  - The number of conformations (`num_conformations`) is chosen by you.  \n",
        "  - Each residue is represented by one bead (coarse-grained representation).\n",
        "\n",
        "\n",
        "### Type of deep generative model\n",
        "idpGAN is a **[Generative Adversarial Network (GAN)](https://en.wikipedia.org/wiki/Generative_adversarial_network)**.\n",
        "\n",
        "GANs were popular in early generative modeling, but are less common today for problems due to training instability and poor performance on complex datasets.\n",
        "\n",
        "### Training data\n",
        "\n",
        "IdpGAN was trained on **coarse-grained (CG) molecular dynamics (MD)** simulations of IDPs.  \n",
        "Their sequences were taken from **[DisProt](https://disprot.org)**.\n",
        "\n",
        "The simulations were performed using **[COCOMO-1](https://pubs.acs.org/doi/abs/10.1021/acs.jctc.2c00856)**.\n",
        "\n",
        "- COCOMO-1 is a one-bead-per-residue CG model.\n",
        "- It is conceptually similar to CALVADOS2 (which you already know). Both aim to reproduce some ensemble properties of IDPs.\n",
        "\n",
        "You can think of idpGAN as a **fast emulator of COCOMO-1 simulations**.\n",
        "\n",
        "‚ùì**Question**‚ùì: since idpGAN was trained on COCOMO-1 simulations:\n",
        "* What ensemble properties should it reproduce well (think about CALVADOS)?\n",
        "* What properties might it fail to capture?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AtSgA4J30VGm"
      },
      "source": [
        "## Load the idpGAN model\n",
        "\n",
        "Loading the machine learning model (e.g.: loading the weights of its neural net) is typically the first step in using them."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ygAAfgZp0EUo"
      },
      "outputs": [],
      "source": [
        "#\n",
        "# Setup the model.\n",
        "#\n",
        "\n",
        "# Set the data directory of the idpGAN Git repository. Do not change if\n",
        "# running on Colab.\n",
        "data_path = \"data\"\n",
        "\n",
        "# Setup the device in which we want to run the PyTorch models.\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "print(\"- Using PyTorch device: '%s'\" % device)\n",
        "\n",
        "# Load the generator network that was used in the idpGAN paper.\n",
        "netg = load_netg_article(model_fp=os.path.join(data_path, \"generator.pt\"),\n",
        "                         device=device)\n",
        "print(\"- IdpGAN generator loaded\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Sd8f6XFM3i7H"
      },
      "source": [
        "# Part 2/6 - Generate a first ensemble\n",
        "\n",
        "Let's generate an IDP ensemble. We will use the [C-terminal part of human prothymosin Œ±](https://www.pnas.org/doi/10.1073/pnas.1001743107) (here named: `protac`, tough it is totally unrelated to [PROTACS](https://en.wikipedia.org/wiki/Proteolysis_targeting_chimera)), which is an intrinsically disordered region (IDR). This sequence was in the test set of idpGAN."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XmVoAino7_XJ"
      },
      "source": [
        "## Generation"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Define the sequence."
      ],
      "metadata": {
        "id": "t8IfNs2tnyBQ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hiQuFlky72zd"
      },
      "outputs": [],
      "source": [
        "# Sequence.\n",
        "seq = \"CEEGGEEEEEEEEGDGEEEDGDEDEEAESATGKRAAEDDEDDDVDTKKQKTDEDC\"\n",
        "seq_name = \"protac\"\n",
        "print(\"sequence length:\", len(seq))\n",
        "\n",
        "# Number of conformations to generate. 1000 is already enough in many applications.\n",
        "num_conformations = 1000"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Actually generate the ensemble."
      ],
      "metadata": {
        "id": "Vl9ao43cn0Hn"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DZUUSd-L8G1z"
      },
      "outputs": [],
      "source": [
        "%%time\n",
        "xyz_gen = netg.predict_idp(n_samples=num_conformations, aa_seq=seq,\n",
        "                           device=device).cpu().numpy()\n",
        "print(\"- the generated xyz is an array of shape:\", xyz_gen.shape)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cPzHjolu8UeY"
      },
      "source": [
        "Congrats! You have generated an IDP ensemble with ML. How much time did it take (see: `Wall time` in the cell above)?\n",
        "\n",
        "‚ùì**Question**‚ùì:\n",
        "  - The output above is an array with a shape like `(num_conformations, 55, 3)`? What is this array representing?\n",
        "\n",
        "**Note**: this is a common way of representing structural ensemble data in computational biology and deep learning."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LmPlYJQ19hfx"
      },
      "source": [
        "## Analysis via IDPET\n",
        "\n",
        "Now, let's analyze the ensemble. The output of idpGAN is a [numpy](https://numpy.org) array. We need to convert it into a topology (PDB) file and a trajectory file first.\n",
        "\n",
        "**Optional note: what are topology and trajectory files?** This combination is one of the most common ways of storing ensemble and Molecular Dynamics data.\n",
        "- **Trajectory file**: a binary file containing all the xyz coordinates of the conformations in your system. It often does not store information about atom types (e.g.: carbon, oxygen, etc...).\n",
        "- **Topology file**: a molecular structure file (here a PDB file) containing the type of atoms of your system. In some tools (idpGAN here and BioEmu) the PDB file also contains the coordinates of the first conformation in the ensemble (this is arbitrary).\n",
        "\n",
        "Once we have that we can load in IDPET. Use the cell below to do it."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ysahXkjL8KIS"
      },
      "outputs": [],
      "source": [
        "# Run this to convert to a topology and PDB files that will be saved on the cloud.\n",
        "to_trajectory_and_top(xyz_gen, seq, f\"idpgan.{seq_name}\", data_path)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4E7GFC4Z_Yub"
      },
      "source": [
        "Now let's load the ensemble in IDPET."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "whMAcHJN_NXr"
      },
      "outputs": [],
      "source": [
        "from idpet.ensemble import Ensemble\n",
        "from idpet.ensemble_analysis import EnsembleAnalysis\n",
        "from idpet.visualization import Visualization\n",
        "\n",
        "\n",
        "# Load the ensemble data.\n",
        "ensemble = Ensemble(\n",
        "    code=seq_name,\n",
        "    data_path=os.path.join(data_path, f\"idpgan.{seq_name}.xtc\"),\n",
        "    top_path=os.path.join(data_path, f\"idpgan.{seq_name}.pdb\")\n",
        ")\n",
        "\n",
        "analysis = EnsembleAnalysis([ensemble])\n",
        "analysis.load_trajectories()\n",
        "# Make the visualization object for visualizing ensemble features.\n",
        "vis = Visualization(analysis=analysis)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JEJVeG-0AG19"
      },
      "source": [
        "How do we analyze an ensemble? The analysis of structural ensembles is an art/science in its own! It depends on the type of data (small vs big proteins, globular vs IDP, MD trajectories vs generic ensemble data).\n",
        "\n",
        "For single chain IDPs, we typically begin by analyzing their [radius of gyration](https://en.wikipedia.org/wiki/Radius_of_gyration) (Rg), which is a measure of chain compactness.\n",
        "\n",
        "Let's plot a the distribution of the radius of gyration of the conformations in our ensemble.\n",
        "\n",
        "‚ùì**Question**‚ùì: given an IDPs and a globular (folded) protein with the same number of amino acids, which one do you expect to have an higher Rg?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uwlii5VH_fsp"
      },
      "outputs": [],
      "source": [
        "vis.radius_of_gyration(violin_plot=True);"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cYSkMciCDxB-"
      },
      "source": [
        "Nice! But right now this is not very informative... Rg mostly makes sense in relative terms (e.g.: the Rg of protein X is larger/smaller the one of protein Y).\n",
        "\n",
        "Later, you will generate ensembles for many IDPs and will study the relationship of Rg with sequence length.\n",
        "\n",
        "Hopefully you will get a better intution of Rg."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-7nfYbCTDLFi"
      },
      "source": [
        "Let's also plot the [contact map](https://en.wikipedia.org/wiki/Protein_contact_map) of the ensemble, another typical analysis tool of IDP ensembles.\n",
        "\n",
        "**Contact maps**: they indicate the probability of two residues (beads) to have a distance below a threshold (8.0 Angstrom is the default in IDPET) in the ensemble."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iPpYwmiUCisa"
      },
      "outputs": [],
      "source": [
        "vis.contact_prob_maps(\n",
        "    avoid_zero_count=True,\n",
        "    log_scale=True,\n",
        "    color='Oranges',\n",
        "    threshold=0.8  # the threshold is expressed in units of nanometers.\n",
        ");"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "huTvtNPyEp1k"
      },
      "source": [
        "‚ùì**Question**‚ùì: why are contact frequencies between residues that are far apart in sequence decreasing in our IDP? Is this something you would expect in a typical folded protein structure?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fgOdtTfvFXkj"
      },
      "source": [
        "# Part 3/6 - Multi-ensemble comparison\n",
        "\n",
        "Let's compare our idpGAN protac ensemble to other ensembles:\n",
        "* **A protac ensemble from COCOMO-1**: idpGAN was trained to replicate COCOMO-1. Let's analyze the protac ensemble from actual COCOMO-1 simulations to check if the model is actually approximating MD.\n",
        "* **poly-ala ensemble**: in COCOMO-1, every protein tends to behave as an IDP. However, different sequences have result in different ensemble properties. Let's use the MD ensemble from a poly-alanine chain of the same length of protac to check if the model can capture sequence-specific behavior.\n",
        "\n",
        "Let's begin a new IDPET analysis with three ensembles."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4U6bNnnchRoE"
      },
      "outputs": [],
      "source": [
        "from idpet.ensemble import Ensemble\n",
        "from idpet.ensemble_analysis import EnsembleAnalysis\n",
        "from idpet.visualization import Visualization\n",
        "\n",
        "\n",
        "# Load the ensemble data (,  and\n",
        "seq_names = [\n",
        "    \"idpgan.protac\",  # idpGAN ensemble of protac.\n",
        "    \"protac\",         # COCOMO1 ensemble of protac.\n",
        "    \"polyala\"         # COCOMO1 ensemble for a poly-alanine sequence of the\n",
        "                      # same length of protac.\n",
        "]\n",
        "\n",
        "ensembles = []\n",
        "for seq_name_i in seq_names:\n",
        "    ensemble = Ensemble(\n",
        "        code=seq_name_i,\n",
        "        data_path=os.path.join(data_path, f\"{seq_name_i}.xtc\"),\n",
        "        top_path=os.path.join(data_path, f\"{seq_name_i}.pdb\")\n",
        "    )\n",
        "    ensembles.append(ensemble)\n",
        "\n",
        "analysis_multi = EnsembleAnalysis(ensembles)\n",
        "analysis_multi.load_trajectories()\n",
        "vis_multi = Visualization(analysis=analysis_multi)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dO5eVHXoGQvf"
      },
      "source": [
        "First, let's plot some statistics about our new ensembles. This is useful to quickly see how many residues and conformations our proteins have."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NPhZSfYHGB1Q"
      },
      "outputs": [],
      "source": [
        "analysis_multi.get_features_summary_dataframe(\n",
        "    selected_features=[\"rg\", \"end_to_end\", \"flory_exponent\"]\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zsjA6TLkIb6l"
      },
      "source": [
        "## IDPET compactness and contact plots"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "68vxEVvnGfMX"
      },
      "source": [
        "Let's visualize Rg distributions."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gma_LVw8GUg2"
      },
      "outputs": [],
      "source": [
        "vis_multi.radius_of_gyration(violin_plot=True);"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F3mUv1K5Gt3H"
      },
      "source": [
        "Let's also visualize the histograms of the end-to-end distances (the distance between the N-term. and C-term. beads of conformations). This is also a polular property analyzed in IDP ensemble analyses."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "b-TX7nhYGjXo"
      },
      "outputs": [],
      "source": [
        "vis_multi.end_to_end_distances(violin_plot=True, summary_stat='mean', rg_norm=False);"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-0Jqmn2GG0ld"
      },
      "source": [
        "Let's also plot the contact maps of the ensembles."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f1Hcafw6Gwbr"
      },
      "outputs": [],
      "source": [
        "vis_multi.contact_prob_maps(\n",
        "    avoid_zero_count=True,\n",
        "    log_scale=True,\n",
        "    color='Blues',\n",
        "    threshold=0.8  # the threshold is expressed in units of nanometers.\n",
        ");"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Kb-3Zl2LHBgR"
      },
      "source": [
        "What do you think about this comparison so far?\n",
        "\n",
        "* ‚ùì**Question 1**‚ùì: is the idpGAN ensemble closely emulating the reference MD ensemble?\n",
        "* ‚ùì**Question 2**‚ùì: are the protac and poly-ala contact maps consistent with the corresponding Rg violin plots? Hint: look at the contact frequencies between the N- and C-term of the proteins.\n",
        "* ‚ùì**Question 3 (important)**‚ùì: why are the contacts maps of polyala and protac so different? Hint: look at the sequence of protac and try to find a physiochemical explanation."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "m3lXRznZHbRJ"
      },
      "source": [
        "## Visualizing 3D structures\n",
        "\n",
        "Finally, let's actually view some 3D structures of these ensembles.\n",
        "\n",
        "Colab does not easily allow to use advanced molecular visualization software. We will use [Py3DMol](https://github.com/avirshup/py3dmol), which is still quite good.\n",
        "\n",
        "Visualizing the 3D structures of your ensembles is crucial in ensemble analysis!\n",
        "\n",
        "**How**: interact with the mouse with the 3d molecule below."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "20E68DRoG6XJ"
      },
      "outputs": [],
      "source": [
        "# change here to visualize other ensembles.\n",
        "ensemble_name = \"idpgan.protac\"\n",
        "# NOTE: change here to visualize another conformation.\n",
        "# set to `conformation = None`, to view different conformations\n",
        "# in an animation.\n",
        "conformation_index = 0\n",
        "\n",
        "\n",
        "sel_ensemble = analysis_multi[ensemble_name]\n",
        "sel_rg = sel_ensemble.get_features(\"rg\")\n",
        "ee_distances = sel_ensemble.get_features(\"end_to_end\")\n",
        "if conformation_index is not None:\n",
        "    conformation = sel_ensemble.trajectory[conformation_index]\n",
        "    print(f\"Rg of conformation {conformation_index}: {sel_rg[conformation_index]:.3f} nm\")\n",
        "    print(f\"End-to-end distance of conformation {conformation_index}: {ee_distances[conformation_index]:.3f} nm\")\n",
        "    print('Blue is N-term. and red is C-term.')\n",
        "else:\n",
        "    conformation = sel_ensemble.trajectory\n",
        "\n",
        "view = show_structure(\n",
        "    snapshot=conformation,\n",
        "    representation='cg',\n",
        "    cmap_name='Spectral_r'\n",
        ")\n",
        "view.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vn6PbBoyM5ZG"
      },
      "source": [
        "‚ùì**Question**‚ùì: are Rg and end-to-end distances correlated? (only for experienced coders: can you make a plot of Rg vs end-to-end of a few conformations?). You can get the two arrays with:\n",
        "```python\n",
        "sel_rg = sel_ensemble.get_features(\"rg\")\n",
        "ee_distances = sel_ensemble.get_features(\"end_to_end\")\n",
        "```\n",
        "and plot them with `plt.scatter(array_1, array_2)` function of matplotlib."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0uGSxR4wNCr6"
      },
      "source": [
        "# Part 4/6 - Generate ensembles for your own sequence\n",
        "\n",
        "Before moving to the final two parts, you may go back to Part 3 and generate an ensemble for **any sequence of your choice**.\n",
        "\n",
        "You will not have a reference MD ensemble for this sequence, the goal is to analyze the ensemble on its own.\n",
        "\n",
        "Generate the ensemble and use the IDPET visualization tools from Part 3 (e.g., `vis.contact_prob_maps`) to characterize it.\n",
        "\n",
        "This section is meant for exploration. The goal is to develop intuition and test the limits of generative models.\n",
        "\n",
        "If you need inspiration, here are some ideas:\n",
        "\n",
        "### üß™ Physiochemically weird sequences of the same length of protac\n",
        "\n",
        "- Try **poly-TRP** (very hydrophobic)  \n",
        "- Try **poly-GLU** or **poly-LYS** (highly charged chains)  \n",
        "- Try sequence patches (e.g., first 50% of the sequence is GLU, last 50% LYS)\n",
        "\n",
        "Before generating the ensemble, pause and predict:\n",
        "\n",
        "- What should Rg look like with respect to the wt protac?\n",
        "- What should the contact map look like?\n",
        "- Does the generated ensemble match your physical intuition?\n",
        "\n",
        "### üß¨ A globular protein sequence\n",
        "\n",
        "Go to UniProt and retrieve the sequence of a globular protein that you know.\n",
        "\n",
        "- Can idpGAN fold it?\n",
        "- What do the generated structures look like?\n",
        "\n",
        "### üìù 3. Introduce mutations to edit the sequence\n",
        "\n",
        "Get the protac sequence and start adding groups of 3-4 point mutations (you can decide which residue to mutate or do it randomly). Repeat for 2-3 cycles.\n",
        "\n",
        "- How many mutations you need to make the ensemble properties significantly different from the ones of the wt sequence?\n",
        "- Can you design an IDP sequence with some desired property (e.g.: strong contacts between residue X and Y)?\n",
        "- **Hint**: if you want to compare plots of different mutants, you can save the plots as files by right clicking on them and download the images."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "teVLG56plZzz"
      },
      "source": [
        "# Part 5/6 - Testing the limits of idpGAN in predicting experimental behavior\n",
        "\n",
        "Looks like that idpGAN can emulate COCOMO-1 behavior, at least under certain conditions.\n",
        "\n",
        "COCOMO-1, was parametrized to reproduce the experimental Rg of IDPs. So idpGAN can be used to obtain ensembles with Rg values that are in reasonable agreement to experimental ones.\n",
        "\n",
        "But is this always true? We need to understand when/how idpGAN fails."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JD2J8CGlvn_f"
      },
      "source": [
        "## Experimental radius-of-gyration data\n",
        "\n",
        "To find out idpGAN limits, let's generate ensembles for a series of IDPs with experimentally-measured Rg values (kindly provided by Giulio Tesei). These values were obtained by techniques like SAXS or FRET.\n",
        "\n",
        "Here we will use a random subset of this data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "b-m-RpIzM460"
      },
      "outputs": [],
      "source": [
        "rg_data = [\n",
        " {'name': 'Nup49_Rg',\n",
        "  'exp_rg': 1.6,\n",
        "  'seq': 'GCQTSRGLFGNNNTNNINNSSSGMNNASAGLFGSKPFA'},\n",
        " {'name': 'DomainV',\n",
        "  'exp_rg': 2.43,\n",
        "  'seq': 'GAMGTAGNKAALLDQIREGAQLKKVEQNSRPVSCSGRDALLDQIRQGIQLKSVADGQESTPPTPAPT'},\n",
        " {'name': 'DSS1',\n",
        "  'exp_rg': 2.5,\n",
        "  'seq': 'MSRAALPSLENLEDDDEFEDFATENWPMKDTELDTGDDTLWENNWDDEDIGDDDFSVQLQAELKKKGVAAC'},\n",
        " {'name': 'HeV_PNT3_CTD_3Y3A',\n",
        "  'exp_rg': 2.73,\n",
        "  'seq': 'MSYYHHHHHHLESTSLYKKAGFTPTEEPPVIPEAAAGSGRRGDLSKSPPRGNVNLDSIKIYTSDDEDENQLEYEDEF'},\n",
        " {'name': 'NUS_Rg',\n",
        "  'exp_rg': 2.5,\n",
        "  'seq': 'GCPSASPAFGANQTPTFGQSQGASQPNPPGFGSISSSTALFPTGSQPAPPTFGTVSSSSQPPVFGQQPSQSAFGSGTTPNFA'},\n",
        " {'name': 'SMAD2_linker',\n",
        "  'exp_rg': 2.99,\n",
        "  'seq': 'GPLPPVLVPRHTEILTELPPLDDYTHSIPENTNFPAGIEPQSNYIPETPPPGYISEDGETSDQQLNQSMDTGSPAELSPTTLSPVNHSLD'},\n",
        " {'name': 'VWF',\n",
        "  'exp_rg': 3.08,\n",
        "  'seq': 'MEDREQAPNLVYMVTGNPASDEIKRLPGDIQVVPIGVGPNANVQELERIGWPNAPILIQDFETLPREAPDLVLQRGGWSHPQFEKGGGSGGGSGGWSHPQFEK'},\n",
        " {'name': 'p27Cv56',\n",
        "  'exp_rg': 2.328,\n",
        "  'seq': 'GSHMKGACGSSVLGTGNPRNQAHVSDTSLEEDDDEQDDSTPDEVSQACTIVASALDINAATPRSPKASPKRKRKRQSTAPAQGNEPPGNAGSVEQTPKKPGLRRRQT'},\n",
        " {'name': 'lambdaN',\n",
        "  'exp_rg': 3.8,\n",
        "  'seq': 'MDAQTRRRERRAEKQAQWKAANPLLVGVSAKPVNRPILSLNRKPKSRVESALNPIDLTVLAEYHKQIESNLQRIERKNQRTWYSKPGERGITCSGRQKIKGKSIPLI'},\n",
        " {'name': 'NUL_Rg',\n",
        "  'exp_rg': 3.0,\n",
        "  'seq': 'GCGFKGFDTSSSSSNSAASSSFKFGVSSSSSGPSQTLTSTGNFKFGDQGGFKIGVSSDSGSINPMSEGFKFSKPIGDFKFGVSSESKPEEVKKDSKNDNFKFGLSSGLSNPVFA'},\n",
        " {'name': 'NHE6cmdd',\n",
        "  'exp_rg': 3.2,\n",
        "  'seq': 'GPPLTTTLPACCGPIARCLTSPQAYENQEQLKDDDSDLILNDGDISLTYGDSTVNTEPATSSAPRRFMGNSSEDALDRELAFGDHELVIRGTRLVLPMDDSEPPLNLLDNTRHGPA'},\n",
        " {'name': 'P7R',\n",
        "  'exp_rg': 2.71,\n",
        "  'seq': 'GSMASASSSQRGRSGRGNFGGGRGGGFGGNDNFGRGGNFSGRGGFGGSRGGGRYGGSGDRYNGFGNDGRNFGGGGSYNDFGNYNNQSSNFGPMKGGNFRGRSSGPYGRGGQYFAKPRNQGGYGGSSSSRSYGSGRRF'},\n",
        " {'name': 'aSyn140',\n",
        "  'exp_rg': 3.55,\n",
        "  'seq': 'MDVFMKGLSKAKEGVVAAAEKTKQGVAEAAGKTKEGVLYVGSKTKEGVVHGVATVAEKTKEQVTNVGGAVVTGVTAVAQKTVEGAGSIAAATGFVKKDQLGKNEEGAPQEGILEDMPVDPDNEAYEMPSEEGYQDYEPEA'},\n",
        " {'name': 'HvASR1',\n",
        "  'exp_rg': 3.46,\n",
        "  'seq': 'GSPEFMAEEKHHHHLFHHKKEGEDFQPAADGGADVYGYSTETVVTGTGNEGEYERITKEEKHHKHKEHLGEMGAVAAGAFALYEKHEAKKDPEHAHKHKIEEELAAAAAVGAGGFVFHEHHEKKQDHKEAKEASGEKKHHLFG'},\n",
        " {'name': 'ANAC046',\n",
        "  'exp_rg': 3.6,\n",
        "  'seq': 'NAPSTTITTTKQLSRIDSLDNIDHLLDFSSLPPLIDPGFLGQPGPSFSGARQQHDLKPVLHHPTTAPVDNTYLPTQALNFPYHSVHNSGSDFGYGAGSGNNNKGMIKLEHSLVSVSQETGLSSDVNTTATPEISSYPMMMNPAMMDGSKSACDGLDDLIFWEDLYTS'},\n",
        " {'name': 'ERNTD_S118D',\n",
        "  'exp_rg': 3.69,\n",
        "  'seq': 'MTMTLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAYEFNAAAAANAQVYGQTGLPYGPGSEAAAFGSNGIGGFPPLNSVSPSPLMLLHPPPQLDPFLQPHGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAKETRY'},\n",
        " {'name': 'CAHSD',\n",
        "  'exp_rg': 4.84,\n",
        "  'seq': 'MSGRNVESHMERNEKVVVNNSGHADVKKQQQQVEHTEFTHTEVKAPLIHPAPPIISTGAAGLAEEIVGQGFTASAARISGGTAEVHLQPSAAMTEEARRDQERYRQEQESIAKQQEREMEKKTEAYRKTAEAEAEKIRKELEKQHARDVEFRKDLIESTIDRQKREVDLEAKMAKRELDREGQLAKEALERSRLATNVEVNFDSAAGHTVSGGTTVSTSDKMEIKRN'},\n",
        " {'name': 'K44',\n",
        "  'exp_rg': 5.2,\n",
        "  'seq': 'MAEPRQEFEVMEDHAGTYGLGDRKDQGGYTMHQDQEGDTDAGLKAEEAGIGDTPSLEDEAAGHVTQARMVSKSKDGTGSDDKKAKGADGKTKIATPRGAAPPGQKGQANATRIPAKTPPAPKTPPSSGEPPKSGDRSGYSSPGSPGTPGSRSRTPSLPTPPTREPKKVAVVRTPPKSPSSAKSRLQTAPVPMPDLKNVKSKIGSTENLKHQPGGGKVQIVYKPVDLSKVTSKCGSLGNIHHKPGGGQVEVKSEKLDFKDRVQSKIGSLDNITHVPGGGNKKIE'},\n",
        " {'name': 'GHRICD',\n",
        "  'exp_rg': 6.0,\n",
        "  'seq': 'SKQQRIKMLILPPVPVPKIKGIDPDLLKEGKLEEVNTILAIHDSYKPEFHSDDSWVEFIELDIDEPDEKTEESDTDRLLSSDHEKSHSNLGVKDGDSGRTSCCEPDILETDFNANDIHEGTSEVAQPQRLKGEADLLCLDQKNQNNSPYHDACPATQQPSVIQAEKNKPQPLPTEGAESTHQAAHIQLSNPSSLSNIDFYAQVSDITPAGSVVLSPGQKNKAGMSQCDMHPEMVSLCQENFLMDNAYFCEADAKKCIPVAPHIKVESHIQPSLNQEDIYITTESLTTAAGRPGTGEHVPGSEMPVPDYTSIHIVQSPQGLILNATALPLPDKEFLSSCGYVSTDQLNKIMP'},\n",
        " {'name': 'ht2N4R',\n",
        "  'exp_rg': 6.7,\n",
        "  'seq': 'MAEPRQEFEVMEDHAGTYGLGDRKDQGGYTMHQDQEGDTDAGLKESPLQTPTEDGSEEPGSETSDAKSTPTAEDVTAPLVDEGAPGKQAAAQPHTEIPEGTTAEEAGIGDTPSLEDEAAGHVTQARMVSKSKDGTGSDDKKAKGADGKTKIATPRGAAPPGQKGQANATRIPAKTPPAPKTPPSSGEPPKSGDRSGYSSPGSPGTPGSRSRTPSLPTPPTREPKKVAVVRTPPKSPSSAKSRLQTAPVPMPDLKNVKSKIGSTENLKHQPGGGKVQIINKKLDLSNVQSKCGSKDNIKHVPGGGSVQIVYKPVDLSKVTSKCGSLGNIHHKPGGGQVEVKSEKLDFKDRVQSKIGSLDNITHVPGGGNKKIETHKLTFRENAKAKTDHGAEIVYKSPVVSGDTSPRHLSNVSSTGSIDMVDSPQLATLADEVSASLAKQGL'}\n",
        "]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pksVqOz0u-hE"
      },
      "source": [
        "## Generate the ensembles\n",
        "\n",
        "Let's start generating idpGAN ensembles for these sequences.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_R5424cIur16"
      },
      "outputs": [],
      "source": [
        "# Helper function, feel free to use. Allows you to compute the Rg of\n",
        "# idpGAN conformations directly on NumPy arrays. In this way, you don't\n",
        "# need to convert them to trajectories and analyze in IDPET.\n",
        "\n",
        "\n",
        "def compute_rg(xyz):\n",
        "    \"\"\"\n",
        "    Takes as input a numpy array of shape (B, L, 3).\n",
        "        B: batch size, L: number of atoms.\n",
        "    Adapted from the mdtraj library: https://github.com/mdtraj/mdtraj.\n",
        "    \"\"\"\n",
        "    num_atoms = xyz.shape[1]\n",
        "    # Assign equal masses to each residue, a tolerable approximation here.\n",
        "    masses = np.ones(num_atoms)\n",
        "    weights = masses / masses.sum()\n",
        "    mu = xyz.mean(1)\n",
        "    centered = (xyz.transpose((1, 0, 2)) - mu).transpose((1, 0, 2))\n",
        "    squared_dists = (centered ** 2).sum(2)\n",
        "    Rg = (squared_dists * weights).sum(1) ** 0.5\n",
        "    return Rg"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "‚ùóÔ∏èÔ∏èÔ∏èÔ∏èÔ∏è**Important note**‚ùóÔ∏è: Note that we generally consider experimental values as ensemble-average values. Therefore, when you generate/simulate an ensemble, you should compare its average Rg value with the experimental one."
      ],
      "metadata": {
        "id": "oFuNWy53K15j"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VHV0J9RGw2yJ"
      },
      "outputs": [],
      "source": [
        "# Will use less conformation here to speed up things.\n",
        "num_conformations_rg = 100\n",
        "idpgan_ensembles = []\n",
        "idpgan_rg = []\n",
        "\n",
        "for s in rg_data:\n",
        "    print(f\"sampling {s['name']} L={len(s['seq'])}\")\n",
        "    xyz_gen_i = netg.predict_idp(\n",
        "        n_samples=num_conformations_rg, aa_seq=s[\"seq\"],\n",
        "        device=device\n",
        "    ).cpu().numpy()\n",
        "    idpgan_ensembles.append(xyz_gen_i)\n",
        "    rg_gen_i = compute_rg(xyz_gen_i)\n",
        "    avg_rg_gen_i = rg_gen_i.mean()\n",
        "    idpgan_rg.append(avg_rg_gen_i)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eO6bVqhEvd5N"
      },
      "source": [
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "‚ùì**Question (important)**‚ùì: why are experimentally-measured values considered ensemble-averages? They are averages over what?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_n7In3ksyNEb"
      },
      "source": [
        "‚ùì**Question (optional)**‚ùì: have you noticed how the speed of sampling is changing with sequence length? Can you generate the ensembles again and print (or plot, if you want) the \"sequence length \" vs \"ensemble generation speed\" relationship? You can time a Python function in your code via:\n",
        "```python\n",
        "import time\n",
        "\n",
        "time_start = time.time()\n",
        "your_python_func()  # call your function here.\n",
        "time_end = time.time()\n",
        "time_it_took = time_end - time_start  # by default, in units of seconds.\n",
        "print(time_it_took)\n",
        "```\n",
        "\n",
        "Is the relationship linear? If not, where might be this coming from (hint: the idpGAN generator is a transformer neural network that has sequences of length `N` as input/output. Try googling \"transformers computational complexity\"). What is the neural network architecture of more modern deep ensemble generators (e.g.: BioEmu)? Do you expect similar \"protein size\" vs \"running time\" relationships?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EezV_aOezTPx"
      },
      "source": [
        "## Comparison with experiments\n",
        "\n",
        "Now, let's plot the \"predictions\" and \"experimental values\" vs \"length\"."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-oRgSPzJzbAo"
      },
      "outputs": [],
      "source": [
        "idp_lengths = [len(r[\"seq\"]) for r in rg_data]\n",
        "exp_rg = [r[\"exp_rg\"] for r in rg_data]\n",
        "\n",
        "plt.scatter(idp_lengths, exp_rg, label=\"experimental\")\n",
        "plt.scatter(idp_lengths, idpgan_rg, label=\"idpGAN\")\n",
        "plt.xlabel(\"length\")\n",
        "plt.ylabel(\"Rg [nm]\")\n",
        "plt.legend()\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "‚ùì**Question:**‚ùì are predictions equally good across the length scale? Or do they get worse in some condition?"
      ],
      "metadata": {
        "id": "GpzfQjDjMAcv"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vyDEAvHC1ZUh"
      },
      "source": [
        "## Checking the physical realism of idpGAN conformations\n",
        "\n",
        "From literature ([COCOMO-1 paper](https://pubs.acs.org/doi/10.1021/acs.jctc.2c00856)) we know that COCOMO-1 tends to overestimate the compactness of longer IDPs. In the paper they wrote:\n",
        "```\n",
        "\"... our model also shows systematic deviations where smaller systems tend to be less compact, whereas larger systems are more compact than suggested by the experimental values.\"\n",
        "```\n",
        "\n",
        "Ok! Is our observation explainable only by the properties of the COCOMO-1 training data? (Indeed: this over-compactness issue was solved in [COCOMO-2](https://pubs.acs.org/doi/10.1021/acs.jctc.4c01460))!\n",
        "\n",
        "idpGAN is just emulating its training data and everything is explained, right?\n",
        "\n",
        "Maybe, but we are interested in truly understanding the behavior our model! Let's investigate further."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jpPjQxuI9tU_"
      },
      "source": [
        "### Inspection of idpGAN conformations\n",
        "\n",
        "One of the weakness of current AI ensemble/structure generators is the physical realism of their conformations: sometimes they contain atom-atom clashes or bond lenth violations, which is quite bad.\n",
        "\n",
        "Let's define some physical errors in COCOMO-1 CG structures:\n",
        "\n",
        "* **Clashes**: any pair of non-bonded (i.e.: non adjacent) beads at a distance: $d < 0.25\\,nm$ (note that a bead in COCOMO-1 represents a residue, so 0.30 nm is a very short distance for two residues).\n",
        "* **Bond length violations**: any pair of bonded (i.e.: adjacent along the primary sequence) beads with a distance outside the range: $0.30\\,nm < d < 0.46\\,nm$.\n",
        "\n",
        "Do idpGAN conformations contain such inaccuracies? Below we define some functions that can be used to evaluate these structural inaccuracies."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "j80TTyi58Yom"
      },
      "outputs": [],
      "source": [
        "#\n",
        "# NOTE: this functions are only valid for CG protein structures. They will not\n",
        "# work with all-atom protein structures (in the last part of the notebook).\n",
        "#\n",
        "\n",
        "def compute_bond_lengths(xyz):\n",
        "    \"\"\"Using numpy vectorization.\"\"\"\n",
        "    # distance vectors between all pairs of adjacent residues. shape: (B, L-1, 3)\n",
        "    delta = xyz[:,:-1,:] - xyz[:,1:,:]\n",
        "    # distances between all pairs of adjacent residues. shape: (B, L-1)\n",
        "    dist = np.sqrt(\n",
        "        # squared norm of the distance vectors.\n",
        "        np.sum(\n",
        "            np.square(delta), axis=-1\n",
        "        )\n",
        "    )\n",
        "    return dist\n",
        "\n",
        "def compute_bond_length_violations_in_ensemble(bond_lengths, min_d=0.30, max_d=0.46):\n",
        "    \"\"\"Compute the average number of violations in an ensemble\"\"\"\n",
        "    # boolean array for labeling violations. shape: (B, L-1)\n",
        "    violations = (min_d > bond_lengths) | (bond_lengths > max_d)\n",
        "    # compute the number of violations of each conformation. shape: (B, )\n",
        "    violations = violations.sum(axis=1)\n",
        "    # compute the average over all conformations.\n",
        "    violations = violations.mean(axis=0)\n",
        "    return violations\n",
        "\n",
        "\n",
        "def compute_nonbonded_distances(xyz):\n",
        "    \"\"\"Using numpy vectorization.\"\"\"\n",
        "    # Get the indices of all pair of non-bonded atoms. shape: [2, num_distances]\n",
        "    indices = np.triu_indices(\n",
        "        xyz.shape[1],\n",
        "        k=2  # exclude pairs that have sequence separation < 2.\n",
        "    )\n",
        "    # distance vectors between all pairs of non-bonded residues. shape: (B, num_distances, 3)\n",
        "    delta = xyz[:,indices[0],:] - xyz[:,indices[1],:]\n",
        "    # distances between all pairs of adjacent residues. shape: (B, L-1)\n",
        "    dist = np.sqrt(\n",
        "        # squared norm of the distance vectors.\n",
        "        np.sum(\n",
        "            np.square(delta), axis=-1\n",
        "        )\n",
        "    )\n",
        "    return dist\n",
        "\n",
        "def compute_clashes_in_ensemble(nonbonded_distances, min_d=0.25):\n",
        "    \"\"\"Compute the average number of clashes in an ensemble\"\"\"\n",
        "    # boolean array for labeling violations. shape: (B, L-1)\n",
        "    violations = (min_d < nonbonded_distances)\n",
        "    # compute the number of clashes of each conformation. shape: (B, )\n",
        "    violations = violations.sum(axis=1)\n",
        "    # compute the average over all conformations.\n",
        "    violations = violations.mean(axis=0)\n",
        "    return violations"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's inspect the 3d structures of some idpGAN conformations:\n",
        "- Let's first analyze conformations for a protein that idpGAN seems to model reasonably well (Nup49_Rg).\n",
        "- Then, find the name of a protein that idpGAN models incorrectly and visualize its conformations.\n",
        "\n",
        "‚ùì**Questions**‚ùì: Can you spot some differences in physical properties (e.g.: clashes)? Also, use the histograms below."
      ],
      "metadata": {
        "id": "O0jZacWhjRLq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# change here to visualize another protein.\n",
        "ensemble_name = \"Nup49_Rg\"\n",
        "# change here to visualize another conformation.\n",
        "# set to `conformation = None`, to view different conformations\n",
        "# in an animation.\n",
        "conformation_index = 0\n",
        "\n",
        "sel_idx = None\n",
        "for i, r in enumerate(rg_data):\n",
        "    if r[\"name\"] == ensemble_name:\n",
        "        sel_idx = i\n",
        "if sel_idx is None:\n",
        "    raise IndexError(f\"No ensemble with name {ensemble_name} was found\")\n",
        "sel_rg_data = rg_data[sel_idx]\n",
        "sel_xyz = idpgan_ensembles[sel_idx]\n",
        "conformation = xyz_to_traj(sel_xyz, sel_rg_data[\"seq\"])\n",
        "\n",
        "if conformation_index is not None:\n",
        "    conformation = conformation[conformation_index]\n",
        "\n",
        "print('Blue is N-term. and red is C-term.')\n",
        "\n",
        "view = show_structure(\n",
        "    snapshot=conformation,\n",
        "    representation='cg',\n",
        "    cmap_name='Spectral_r'\n",
        ")\n",
        "view.show()"
      ],
      "metadata": {
        "id": "lqLW5ckQjR54"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Can you spot outliers that violate the thresholds in the histograms below?"
      ],
      "metadata": {
        "id": "3CZCzVkTmu05"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "bond_lengths = compute_bond_lengths(sel_xyz).ravel()\n",
        "\n",
        "plt.title(f\"bond lengths for {ensemble_name}\")\n",
        "plt.xlabel(\"bond length [nm]\")\n",
        "plt.ylabel(\"density\")\n",
        "plt.hist(bond_lengths, bins=50)\n",
        "plt.axvline(x=0.30, color=\"C3\", label=\"thresholds\")\n",
        "plt.axvline(x=0.46, color=\"C3\")\n",
        "plt.legend()\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "VrD0o1VrkDuq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "nonbonded_distances = compute_nonbonded_distances(sel_xyz).ravel()\n",
        "\n",
        "plt.title(f\"non-bonded distances for {ensemble_name}\")\n",
        "plt.xlabel(\"distance [nm]\")\n",
        "plt.ylabel(\"density\")\n",
        "plt.hist(nonbonded_distances, bins=50)\n",
        "plt.axvline(x=0.25, color=\"C3\", label=\"threshold\")\n",
        "plt.legend()\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "0F5PzUnMjN2o"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_FqlGgVo-zn-"
      },
      "source": [
        "### Sequence length vs physical inaccuracies\n",
        "\n",
        "Finally, compute our scores for **all proteins** and plot our results as a function of sequence length.\n",
        "\n",
        "Ô∏è‚ùóÔ∏èÔ∏è**Note**Ô∏è‚ùóÔ∏è: there are some missing parts in this code! Can you fill-in the gaps?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CJ5P6uyY-3HN"
      },
      "outputs": [],
      "source": [
        "idpgan_bond_length_viol = []\n",
        "idpgan_clashes = []\n",
        "\n",
        "for i in range(len(rg_data)):\n",
        "\n",
        "    print(f\"analyzing {rg_data[i]['name']} L={len(rg_data[i]['seq'])}\")\n",
        "\n",
        "    xyz = idpgan_ensembles[i]\n",
        "    bond_lengths = None  # MISSING PART: Use the correct function here.\n",
        "    bond_length_violations = None  # MISSING PART: Use the correct function here.\n",
        "\n",
        "    nonbonded_distances = None  # MISSING PART: Use the correct function here.\n",
        "    clashes = None  # MISSING PART: Use the correct function here.\n",
        "\n",
        "    if any(\n",
        "        [\n",
        "            a is None for a in \\\n",
        "            [bond_lengths, bond_length_violations, nonbonded_distances, clashes]\n",
        "        ]\n",
        "    ):\n",
        "        raise ValueError(\"You need to add all the correct function from the cell above!\")\n",
        "\n",
        "    idpgan_bond_length_viol.append(bond_length_violations)\n",
        "    idpgan_clashes.append(clashes)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "worYYxJy_rur"
      },
      "source": [
        "Plot the \"length\" vs \"bond length violations\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8BU4e3_S_vQZ"
      },
      "outputs": [],
      "source": [
        "idp_lengths = [len(r[\"seq\"]) for r in rg_data]\n",
        "plt.scatter(idp_lengths, idpgan_bond_length_viol)\n",
        "plt.xlabel(\"length\")\n",
        "plt.ylabel(\"avg. of bond length violations in a conformation\")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "V-IpMzKZ_0fO"
      },
      "source": [
        "And finally the clashes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QfYjiT1f_yLE"
      },
      "outputs": [],
      "source": [
        "idp_lengths = [len(r[\"seq\"]) for r in rg_data]\n",
        "\n",
        "plt.scatter(idp_lengths, idpgan_clashes)\n",
        "plt.xlabel(\"length\")\n",
        "plt.ylabel(\"ensemble_average[clashes in a conformation]\")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hICK_aQXAuMA"
      },
      "source": [
        "‚ùì**Question**‚ùì: what is your final assessment? Is physical quality deteriorating with increasing lengths?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6Ws0eeoqAzSI"
      },
      "source": [
        "## Conclusions\n",
        "\n",
        "Congrats! You broke idpGAN and have correctly diagnosed limits of a deep generative model.\n",
        "\n",
        "An MD force field has strongly repulsive forces for non-bonded atoms and a strong bond length potential, it could never results in such unphysical conformations!\n",
        "\n",
        "Why is idpGAN doing such a poor job on IDPs longer than ~100 residues? By looking at the [idpGAN paper](https://www.nature.com/articles/s41467-023-36443-x) you can see the maximum effective length was 120 residues. This explains a lot. It has never seen proteins longer than this.\n",
        "\n",
        "Also, its neural network and training procedure were not optimal (with a small neural network and an heuristic GAN training procedure for handling variable sequence lengths).\n",
        "\n",
        "Therefore the extrapolation ability of the model is limited.\n",
        "\n",
        "Modern ensemble generators have more robust training and extrapolation abilities, with vastly improved physical realism (see below).\n",
        "\n",
        "But they may also have limits related to training data (e.g.: length, disorder fraction, sequence features, domain architecture, etc...). For this reason, always analyze your data carefully!"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Jm0mIxbsA4RU"
      },
      "source": [
        "# Part 6/6 - Atomistic ensembles and last-generation deep ensemble generators\n",
        "\n",
        "## More recent ML ensemble generators\n",
        "\n",
        "Finally, we are going to analyze some atomistic IDP ensembles from more recent ML ensemble generator.\n",
        "\n",
        "The goal is to introduce you to the analysis of atomistic IDP ensembles and compare modern ML generators to idpGAN.\n",
        "\n",
        "Unfortunately, running these modern methods is computationally expensive and cannot be easily done in a single Colab notebook. You will have to download pre-generated ensembles from our GitHub repo.\n",
        "\n",
        "## Test case\n",
        "We will analyze different ensembles of [ACTR](https://pubs.acs.org/doi/10.1021/ja4045532) a 69-residue IDP (it is often found in IDP ensemble benchmarks).\n",
        "\n",
        "## Ensemble generation methods\n",
        "\n",
        "Here is the list of ensembles we will analyze:\n",
        "- **Reweighted atomistic MD (PED00536e001)**: this is an ensemble deposited on [PED](https://proteinensemble.org/entries/PED00536). It was recently constructed by *Borthakur et al.* by running [extensive MD simulations followed by reweighting against experimental data](https://www.nature.com/articles/s41467-025-64098-3#ref-CR53). It was found to be in excellent agreement with the experimental data (NMR and SAXS), we can consider it an high-quality ensemble.\n",
        "- **idpGAN (idpgan-aa)**: an idpGAN ensemble, all-atom details were reconstructed in a post-processing step using the [cg2all neural network](https://www.sciencedirect.com/science/article/pii/S096921262300401X).\n",
        "- **BioEmu (bioemu)**: an ensemble generated by the [BioEmu](https://www.science.org/doi/10.1126/science.adv9817) AI ensemble generator, using its default parameters. BioEmu was not trained on IDPs, but [was reported](https://www.biorxiv.org/content/10.1101/2025.10.18.680935v3) to generate IDP ensembles that are in somewhat good agreement with experimental data.\n",
        "- **Boltz-2 default (boltz2)**: ensemble generated by running [Boltz-2](https://www.biorxiv.org/content/10.1101/2025.06.14.659707v1) with default parameters, using as input a multiple sequence alignment (MSA). Boltz-2 has not been trained on IDPs. Running Boltz-2 in this modality [reportedly](https://www.biorxiv.org/content/10.1101/2025.10.18.680935v3) does not produce IDP ensembles with good agreement with experimental data.\n",
        "\n",
        "## Goal\n",
        "\n",
        "We know that the PED00536e001 ensemble is in good agreement with experimental data. You goal is to see if the ensembles generated by AI methods can recapitulate its main structural properties."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download the ML-generated ensembles\n",
        "\n",
        "#@markdown Run this cell to download the ensembles\n",
        "\n",
        "data_path = \"data\"\n",
        "\n",
        "# download PED ensemble. The original ensemble contains a few extra-atoms, we remove them here.\n",
        "ped_code = \"PED00536\"\n",
        "!wget https://deposition.proteinensemble.org/api/v1/entries/{ped_code}/ensembles/e001/ensemble-pdb\n",
        "!tar -xzvf ensemble-pdb\n",
        "!rm ensemble-pdb\n",
        "with open(\"filter.txt\", \"w\") as o_fh:\n",
        "    o_fh.write(\"\"\"ATOM   1069  H   GLY A   1\n",
        "ATOM   1070  HE2 HIS A  36\n",
        "ATOM   1071  N   GLY A   1\n",
        "ATOM   1072  H   GLY A   1\"\"\")\n",
        "!grep -v -f filter.txt pdbfile.pdb | sed '/^MODEL *250/,$d' > output.pdb\n",
        "!rm pdbfile.pdb\n",
        "_traj = mdtraj.load(\"output.pdb\")\n",
        "_traj.save(f\"{ped_code}e001.xtc\")\n",
        "_traj[0].save(f\"{ped_code}e001.pdb\")\n",
        "!cp {ped_code}e001.* {data_path}/\n",
        "!rm output.pdb\n",
        "\n",
        "# idpGAN ensembles\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/idpgan.aa.actr.pdb -P $data_path\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/idpgan.aa.actr.xtc -P $data_path\n",
        "# BioEmu\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/bioemu.dpm.pdb -P $data_path\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/bioemu.dpm.xtc -P $data_path\n",
        "# Boltz-2\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/boltz2.msa.pdb -P $data_path\n",
        "!wget https://github.com/giacomo-janson/ml4ngp_training_school_2026/raw/refs/heads/main/data/ensembles/boltz2.msa.xtc -P $data_path\n"
      ],
      "metadata": {
        "id": "Xh5r4Q9KCxIZ",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Initialize the IDPET analysis\n",
        "\n",
        "Run the cell below to initialize the analysis.\n",
        "\n",
        "**Note**: IDPET will have to download and setup the PED00536e001 ensemble, which might take ~1-2 minutes."
      ],
      "metadata": {
        "id": "WzOQotyJDB0o"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dsvnSel_BAKh"
      },
      "outputs": [],
      "source": [
        "from idpet.ensemble import Ensemble\n",
        "from idpet.ensemble_analysis import EnsembleAnalysis\n",
        "from idpet.visualization import Visualization\n",
        "\n",
        "\n",
        "aa_ensembles = [\n",
        "    # This ensemble that can be downloaded from PED via IDPET.\n",
        "    Ensemble(\n",
        "        code='PED00536e001',\n",
        "        data_path=os.path.join(data_path, f\"PED00536e001.xtc\"),\n",
        "        top_path=os.path.join(data_path, f\"PED00536e001.pdb\")\n",
        "    ),\n",
        "\n",
        "    # We will load the other ensembles by loading their topology+trajectory files\n",
        "    # that we downloaded above.\n",
        "    Ensemble(\n",
        "        code=\"idpgan-aa\",\n",
        "        data_path=os.path.join(data_path, f\"idpgan.aa.actr.xtc\"),\n",
        "        top_path=os.path.join(data_path, f\"idpgan.aa.actr.pdb\")\n",
        "    ),\n",
        "    Ensemble(\n",
        "        code=\"bioemu\",\n",
        "        data_path=os.path.join(data_path, f\"bioemu.dpm.xtc\"),\n",
        "        top_path=os.path.join(data_path, f\"bioemu.dpm.pdb\")\n",
        "    ),\n",
        "    Ensemble(\n",
        "        code=\"boltz2\",\n",
        "        data_path=os.path.join(data_path, f\"boltz2.msa.xtc\"),\n",
        "        top_path=os.path.join(data_path, f\"boltz2.msa.pdb\")\n",
        "    )\n",
        "\n",
        "]\n",
        "aa_analysis = EnsembleAnalysis(\n",
        "    aa_ensembles,\n",
        ")\n",
        "aa_analysis.load_trajectories();\n",
        "aa_analysis.random_sample_trajectories(sample_size=125);\n",
        "aa_vis = Visualization(analysis=aa_analysis)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "aa_analysis.get_features_summary_dataframe(\n",
        "    selected_features=[\"rg\", \"end_to_end\", \"flory_exponent\"]\n",
        ")"
      ],
      "metadata": {
        "id": "GjoHaW1uPRQU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Compactness and contact analysis (global features)\n",
        "\n",
        "Let's first visualize the Rg of the ensembles."
      ],
      "metadata": {
        "id": "z_rlvLkuD-H4"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_2X1GUSXqHAT"
      },
      "outputs": [],
      "source": [
        "actr_experimental_rg = 2.63\n",
        "\n",
        "ax = aa_vis.radius_of_gyration(violin_plot=True)\n",
        "ax.axhline(y=actr_experimental_rg,\n",
        "           linestyle=\"--\", color=\"gray\",\n",
        "           label=\"experimental value (SAXS)\")\n",
        "ax.legend()\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's also plot the contact maps of the ensembles."
      ],
      "metadata": {
        "id": "FBJw-cFhOZr2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "aa_vis.contact_prob_maps(\n",
        "    avoid_zero_count=True,\n",
        "    log_scale=True,\n",
        "    color='Blues',\n",
        "    threshold=0.8  # the threshold is expressed in units of nanometers.\n",
        ");"
      ],
      "metadata": {
        "id": "UXy83X5AOfNK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "‚ùì**Questions**‚ùì:\n",
        "- Based on the Rg distribution and contact map patterns, are the ensembles similar to each other?\n",
        "- In the Rg plot, we also included the [experimental Rg value for this protein](https://onlinelibrary.wiley.com/doi/abs/10.1002/pro.435). Which ensemble appears to be in best agreement?"
      ],
      "metadata": {
        "id": "1_XTZbAVbz_6"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Secondary structure analysis (local features)\n",
        "\n",
        "According to experiment, ACTR is reported to form partially stable alpha helices. Are there any helices in our ensembles?\n",
        "\n",
        "The plot below reports the frequency of a residue to be in helical state in an ensemble (0: never helix; 1: always helix)."
      ],
      "metadata": {
        "id": "uq_p_CWudYY5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "aa_vis.relative_dssp_content(\n",
        "    dssp_code='H',  # H: analyze helicity.\n",
        "    figsize=(6,3)\n",
        ");"
      ],
      "metadata": {
        "id": "kBtWpMd_OfLH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The PED00536e001 ensemble was reported to have excellent agreement with NMR data that allows to quantify backbone torsion preference experimentally. So, its helicity is probably a good representation of the experimental behavior.\n",
        "\n",
        "‚ùì**Questions**‚ùì:\n",
        "- how are the deep generative models capturing the helicity? Are they capturing the correct helical regions and helix abundance?\n",
        "- How is idpGAN doing with helicity? Is it producing any? If not, why?"
      ],
      "metadata": {
        "id": "97b6x1QOebKc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Structural analysis exploration\n",
        "\n",
        "Let's visualize some actual 3D structures. At a first glance, do the AI-generated structures look like the MD ones from PED00536e001?"
      ],
      "metadata": {
        "id": "Zs6R9TDugEtZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"- Use the following codes below:\", aa_analysis.ens_codes)"
      ],
      "metadata": {
        "id": "iJWVQb_vhHyl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ensemble_name = \"PED00536e001\"      # change here to visualize other ensembles.\n",
        "conformation_index = 0             # change here to visualize another conformation.\n",
        "\n",
        "sel_ensemble = aa_analysis[ensemble_name]\n",
        "if conformation_index is not None:\n",
        "    conformation = sel_ensemble.trajectory[conformation_index]\n",
        "    sel_rg = sel_ensemble.get_features(\"rg\")\n",
        "    ee_distances = sel_ensemble.get_features(\"end_to_end\")\n",
        "    print(f\"Rg of conformation {conformation_index}: {sel_rg[conformation_index]:.3f} nm\")\n",
        "    print(f\"End-to-end distance of conformation {conformation_index}: {ee_distances[conformation_index]:.3f} nm\")\n",
        "else:\n",
        "    conformation = sel_ensemble.trajectory\n",
        "\n",
        "print('Blue is N-term. and red is C-term.')\n",
        "\n",
        "view = show_structure(\n",
        "    snapshot=conformation,\n",
        "    representation='aa',\n",
        "    cmap_name='Spectral_r'\n",
        ")\n",
        "view.show()"
      ],
      "metadata": {
        "id": "0IaJXRD2OfIt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "‚ùì**Question**‚ùì:\n",
        "- do BioEmu structures contain all atoms? Or are they missing some? Here is the ACTR sequence: `GTQNRPLLRNSLDDLVGPPSNLEGQSDERALLDQLHTLLSNTDATGLEEIDRALGIPELVNQGQALEPKQD`.\n",
        "- is idpGAN showing any atomic clashes?\n",
        "- how do the Boltz-2 conformations (which have a low Rg) look like?"
      ],
      "metadata": {
        "id": "GwCFpoa7hsJH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Conclusions\n",
        "A complete evaluate of these ensembles will involve comparison with extensive experimental data (SAXS, many NMR observables), which is beyond the scope of this notebook.\n",
        "\n",
        "However, based on your analysis:\n",
        "- Which ensemble better captures the experimental Rg?\n",
        "- Which ML ensemble better captures the helicity observed in the MD ensemble?\n",
        "- Is there an ML ensemble which fully captures Rg and helicity at the same time?\n",
        "- Can you rationalize the properties of these ML ensembles based on their training data?"
      ],
      "metadata": {
        "id": "_78SqTIY7DsT"
      }
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "0ZVy8w-kL_Ub"
      ],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
